### PR TITLE
Fix for #2691

### PR DESCRIPTION
**Pull request by searhein**
Hopefully, this fixes the issue #2691. If KOKKOS_HAVE_SERIAL is disabled, `KokkosOpenMPWrapperNode` is used instead of `KokkosSerialWrapperNode`.

### CODE REVIEW

File: packages/shylu/shylu_dd/frosch/test/TestInterfaceSets/main.cpp [hunk 1]
```diff
@@ -65,7 +65,11 @@ typedef unsigned UN;
 typedef double SC;
 typedef int LO;
 typedef int GO;
+#ifdef KOKKOS_HAVE_SERIAL
```
Comment (mhoemmen) [inline packages/shylu/shylu_dd/frosch/test/TestInterfaceSets/main.cpp]: Use `KOKKOS_ENABLE_SERIAL` not `KOKKOS_HAVE_SERIAL`.  The latter is deprecated and will be removed at the next Kokkos develop->master promotion.

File: packages/shylu/shylu_dd/frosch/test/TestLaplacian_Epetra/main.cpp [hunk 2]
```diff
@@ -67,7 +67,11 @@ typedef unsigned UN;
 typedef double SC;
 typedef int LO;
 typedef int GO;
-typedef Kokkos::Compat::KokkosSerialWrapperNode EpetraNode; // Hier Default verwenden???
+#ifdef KOKKOS_HAVE_SERIAL
```
Comment (mhoemmen) [inline packages/shylu/shylu_dd/frosch/test/TestLaplacian_Epetra/main.cpp]: See above comment.

File: packages/shylu/shylu_dd/frosch/test/TestLaplacian_Epetra/main.cpp [hunk 3]
```diff
@@ -67,7 +67,11 @@ typedef unsigned UN;
...
+typedef Kokkos::Compat::KokkosSerialWrapperNode EpetraNode;
+#else
+typedef Kokkos::Compat::KokkosOpenMPWrapperNode EpetraNode;
+#endif
 typedef EpetraNode NO;
 
 using namespace std;
```
Comment (mhoemmen) [inline packages/shylu/shylu_dd/frosch/test/TestLaplacian_Epetra/main.cpp]: It's generally bad practice to do `using namespace std;` at the global scope.  It's acceptable for a unit test .cpp file, but never OK for a header file.

**Comment by searhein** [comment]
@mhoemmen Thanks for you comments. I have modified the code accordingly. Is this ok now?

File: packages/shylu/shylu_dd/frosch/test/TestInterfaceSets/main.cpp [hunk 4]
```diff
@@ -65,16 +65,20 @@ typedef unsigned UN;
 typedef double SC;
 typedef int LO;
 typedef int GO;
+#ifdef KOKKOS_ENABLE_SERIAL
 typedef Kokkos::Compat::KokkosSerialWrapperNode EpetraNode;
+#else
+typedef Kokkos::Compat::KokkosOpenMPWrapperNode EpetraNode;
```
Comment (mhoemmen) [inline packages/shylu/shylu_dd/frosch/test/TestInterfaceSets/main.cpp]: Needs `KOKKOS_ENABLE_OPENMP` protection here.  For example, it's possible to build with OpenMP off, Serial on, and Cuda on.

File: packages/shylu/shylu_dd/frosch/test/TestLaplacian_Epetra/main.cpp [hunk 5]
```diff
@@ -67,17 +67,21 @@ typedef unsigned UN;
 typedef double SC;
 typedef int LO;
 typedef int GO;
-typedef Kokkos::Compat::KokkosSerialWrapperNode EpetraNode; // Hier Default verwenden???
+#ifdef KOKKOS_ENABLE_SERIAL
+typedef Kokkos::Compat::KokkosSerialWrapperNode EpetraNode;
+#else
```
Comment (mhoemmen) [inline packages/shylu/shylu_dd/frosch/test/TestLaplacian_Epetra/main.cpp]: Needs `KOKKOS_ENABLE_OPENMP` protection here.

**Comment by mhoemmen** [comment]
@searhein I should actually not have approved these changes since I'm not sure if this will break the CUDA build.  Would you happen to know if this works in a CUDA build with OpenMP disabled?

**Comment by searhein** [comment]
@mhoemmen I thought Epetra does not work with CUDA anyway, right? The the issue was just for the Epetra test, there were no problems with the Tpetra test.

**Comment by searhein** [comment]
This test is explicitly using Epetra. If Epetra is not enabled, I don't want the test to compile at all. 
There is another equivalent test which uses Tpetra instead.

**Comment by bartlettroscoe** [comment]
I tested this branch merged into the current 'develop' branch with the GCC 4.8.4 + OpenMPI 1.10.1 + OpenMP build as in the "Steps to Reproduce" section at the top of #2691 and these three tests still fail:

```
Test project /ascldap/users/rabartl/Trilinos.base/BUILDS/GCC-4.8.4/MPI_RELEASE_DEBUG_SHARED_PT_OPENMP
    Start 4: ShyLU_DDFROSch_test_frosch_laplacian_tpetra_2d_gdsw_MPI_4
    Start 5: ShyLU_DDFROSch_test_frosch_laplacian_tpetra_2d_rgdsw_MPI_4
    Start 1: ShyLU_DDFROSch_test_frosch_laplacian_epetra_2d_gdsw_MPI_4
    Start 2: ShyLU_DDFROSch_test_frosch_laplacian_epetra_2d_rgdsw_MPI_4
1/5 Test #5: ShyLU_DDFROSch_test_frosch_laplacian_tpetra_2d_rgdsw_MPI_4 ...   Passed    1.40 sec
    Start 3: ShyLU_DDFROSch_test_frosch_interfacesets_2D_MPI_4
2/5 Test #4: ShyLU_DDFROSch_test_frosch_laplacian_tpetra_2d_gdsw_MPI_4 ....   Passed    1.42 sec
3/5 Test #2: ShyLU_DDFROSch_test_frosch_laplacian_epetra_2d_rgdsw_MPI_4 ...***Failed    1.74 sec
4/5 Test #1: ShyLU_DDFROSch_test_frosch_laplacian_epetra_2d_gdsw_MPI_4 ....***Failed    1.74 sec
5/5 Test #3: ShyLU_DDFROSch_test_frosch_interfacesets_2D_MPI_4 ............***Failed    1.53 sec

40% tests passed, 3 tests failed out of 5

Subproject Time Summary:
ShyLU_DD    =  31.28 sec*proc (5 tests)

Total Test time (real) =   2.96 sec

The following tests FAILED:
          1 - ShyLU_DDFROSch_test_frosch_laplacian_epetra_2d_gdsw_MPI_4 (Failed)
          2 - ShyLU_DDFROSch_test_frosch_laplacian_epetra_2d_rgdsw_MPI_4 (Failed)
          3 - ShyLU_DDFROSch_test_frosch_interfacesets_2D_MPI_4 (Failed)
```

I will submit to CDash for more data.

**Comment by bartlettroscoe** [comment]
I pushed the merge between this PR branch and 'develop' to:

* https://github.com/bartlettroscoe/Trilinos/tree/2961-shylu-fix

That seems to show that these new commits are included:

* https://github.com/trilinos/Trilinos/compare/develop...bartlettroscoe:2961-shylu-fix

I then submitted results to CDash shown at:

* https://testing-vm.sandia.gov/cdash/index.php?project=Trilinos&parentid=3522649

This showed this tested the version:

```
Trilinos repos versions:
--------------------------------------------------------------------------------
*** Base Git Repo: Trilinos
51843db [Thu May 17 06:36:24 2018 -0600] <rabartl@sandia.gov>
Merge remote-tracking branch 'github/develop' into 2961-shylu-fix
 --------------------------------------------------------------------------------
```

You can see the three failing tests here:

* https://testing-vm.sandia.gov/cdash/viewTest.php?onlyfailed&buildid=3522650

The failing test output shows:

```
terminate called after throwing an instance of 'Xpetra::Exceptions::RuntimeError'
  what():  /ascldap/users/rabartl/Trilinos.base/Trilinos/packages/xpetra/src/CrsMatrix/Xpetra_EpetraCrsMatrix.hpp:222:

Throw number = 1

Throw test that evaluated to true: true

Xpetra::EpetraCrsMatrix only available for GO=int or GO=long long with EpetraNode (Serial or OpenMP depending on configuration)
```

Any idea what the problem might be?

**Comment by mhoemmen** [comment]
@searhein wrote:
> This test is explicitly using Epetra. If Epetra is not enabled, I don't want the test to compile at all.

Sorry, it's possible I didn't explain quite right -- it is possible to do all of the following:

  - Enable Epetra
  - Disable OpenMP
  - Enable CUDA

Whether or not Kokkos enables Kokkos, OpenMP, and/or Cuda, affects whether the corresponding `*WrapperNode` class exists.

I don't mean to make you do work; I just want to help you learn about all the issues that may arise when you write code that refers to different Kokkos things.  Thanks!

**Comment by srajama1** [comment]
@ndellingwood : Is there a way to confirm that this PR fixes the ATDM problem, then we can always have this test enabled.

**Comment by ndellingwood** [comment]
@srajama1 I think we can revert Ross's commit that disabled the test, then rerun using the atdm scripts on the test bed where this failed. I'll have to dig back through the issues to track down the instructions, I'm not on my work computer now but can do this checked when I'm in front of the computer for a spurt.

**Pull request closed without merging**